### PR TITLE
[WIP] ReactNativeFiber mixins refactor

### PR DESCRIPTION
**Pull request by bvaughn**
**Note** This PR has changed significantly since it was initially created. (See commit history for more.)

Addressed a circular dependency between `NativeMethodsMixin` and `ReactNativeFiber`:
* Created new `ReactNativeFiberHostComponent` with similar interface but without unnecessary call to `findNodeHandle`.
* Created new Flow interface for mixins to ensure `ReactNativeFiberHostComponent` and `NativeMethodsMixinUtils` stay synced.
* Moved helper methods to `NativeMethodsMixinUtils` to be shared between the 2 wrappers.

Fixed bug in `createReactNativeComponentClass` that allowed duplicate view names to be registered and added a unit test to guard against a regression.

### CODE REVIEW

**Comment by sebmarkbage** [comment]
We actually have a task to undo all the tests that reach into internals like this. :) It makes refactoring harder.

You can express this using public APIs. You can test the first two by calling `createReactNativeComponentClass`.

You can test the last one by rendering a view like `<foo />` using the renderer.

**Comment by bvaughn** [comment]
Yeah, I considered testing with `createReactNativeComponentClass` but the behavior varies between fiber and stack. (Stack doesn't error.)

**Comment by sebmarkbage** [comment]
That's an argument for doing it so that we're aware of the differences. It might actually be a problem for us because people copy paste stuff into their own modules. Not guaranteed to have unique names. Would be better with an incremental counter.

**Comment by sebmarkbage** [comment]
I suppose it's unlikely that you have two different modules targeting the same native uiViewClassName but it's possible.

**Comment by bvaughn** [comment]
> I don't think this is the strategy we should go with at all though since it can break the ecosystem when people load different modules.

Yes, I agree. We could switch to some other uid (like a counter) to make this particular issue more robust, but there's also the related issue of native mixins so maybe this is a better time to re-think the return value of `ReactNativeViewConfigRegistry.register`

Let's sync up tomorrow at the office and talk more.

File: src/renderers/shared/fiber/ReactFiber.js [hunk 1]
```diff
@@ -338,10 +338,16 @@ function createFiberFromElementType(
 ) : Fiber {
   let fiber;
   if (typeof type === 'function') {
-    fiber = shouldConstruct(type) ?
-      createFiber(ClassComponent, key) :
-      createFiber(IndeterminateComponent, key);
-    fiber.type = type;
+    // @TODO (bvaughn) This is temporary hack just to get things working.
+    if (typeof type.__reactInternalHostComponentFlag !== 'undefined') {
+      fiber = createFiber(HostComponent, key);
```
Comment (gaearon) [inline src/renderers/shared/fiber/ReactFiber.js]: I hope this doesn't stay for a couple of years like similar hacks have stayed in Stack :-)

Comment (gaearon) [inline src/renderers/shared/fiber/ReactFiber.js]: What is the plan for unhacking this?

Comment (bvaughn) [inline src/renderers/shared/fiber/ReactFiber.js]: This PR is just me poking things a bit and familiarizing myself with the problem. Sebastian and I are going to talk more about it when he gets in this morning. 😄

**Comment by janicduplessis** [comment]
There was a circular dependency that was causing a crash when running the Native Animated example in UIExplorer with Fiber, ReactNative -> NativeModulesMixin -> ReactNative. I was about to send a fix but I _think_ this PR should also fix it.

**Comment by bvaughn** [comment]
I've moved this effort to an internal diff for now because it's easier to test its impact that way. I'm going to close this PR (since it's changed so much from the initial commit anyway) and I'll open a new one once the internal diff has been merged. 😄

**Comment by gaearon** [comment]
>I'll open a new one once the internal diff has been merged

Wouldn't it just show up as a commit then? I don't think you need to do anything once internal diff lands.

**Comment by gaearon** [comment]
Oh, we're in React repo. Nevermind 😄

**Pull request closed without merging**